pkg/arvo/app/group-store.hoon

```diff
@@ -85,85 +130,322 @@
   --
 ::
 |_  bol=bowl:gall
+++  ship-rank
+  |=  =ship
+  ^-  rank
+  =/  size
+    (met 3 ship)
+  ?:  (lte size 2)
+    %0
+  ?:  (lte size 4)
+    %1
+  ?:  (lte size 8)
+    %2
+  %3
+++  peek-group
+  |=  =group-id
+  ^-  (unit group)
+  (~(get by groups) group-id)
+
+++  peek-group-join
+  |=  [=group-id =ship]
+  =/  =group
+    (~(gut by groups) group-id *group)
+  =*  policy  policy.group
+  ?-  -.policy
+      %invite
+    |((~(has in pending.policy) ship) (~(has in members.group) ship))
+      %open
+    &(!(~(has in banned.policy) ship) (~(has in ranks.policy) (ship-rank ship)))
+  ==
+++  peek-group-role
+  |=  [=group-id =ship]
+  =/  =group
+    (~(got by groups) group-id)
+  =*  policy  policy.group
+  =*  tag-queries  tag-queries.group
+  =/  admins=(set ^ship)
+    (~(gut by tag-queries) %admin ~)
+  ?:  (~(has in admins) ship)
+    `%admin
+  =/  mods
+    (~(gut by tag-queries) %moderator ~)
+  ?:  (~(has in mods) ship)
+    `%moderator
+  =/  janitors
+    (~(gut by tag-queries) %janitor ~)
+  ?.  (~(has in janitors) ship)
+    ~
+  `%janitor
 ::
-++  poke-group-action
-  |=  action=group-action
+++  poke-group-update
+  |=  =update:store
   ^-  (quip card _state)
   ?>  (team:title our.bol src.bol)
-  ?-  -.action
-      %add       (handle-add action)
-      %remove    (handle-remove action)
-      %bundle    (handle-bundle action)
-      %unbundle  (handle-unbundle action)
+  |^
+  ?-  -.update
+      %add-group       (add-group +.update)
+      %add-members     (add-members +.update)
+      %remove-members  (remove-members +.update)
+      %add-tag         (add-tag +.update)
+      %remove-tag      (remove-tag +.update)
+      %change-policy   (change-policy +.update)
+      %remove-group    (remove-group +.update)
+      %initial-group   (initial-group +.update)
+      %initial         [~ state]
   ==
+  ::  +add-group: add group to store
+  ::
+  ::    always include ship in own groups, no-op if group already exists
+  ::
+  ++  add-group
+    |=  [=group-id =policy]
+    ^-  (quip card _state)
+    ?:  (~(has by groups) group-id)
+      [~ state]
+    =|  =group
+    =.  members.group
+      (~(put in members.group) our.bol)
+    =.  policy.group   policy
+    =.  groups
+      (~(put by groups) group-id group)
+    :_  state
+    (send-diff %add-group group-id policy)
+  ::  +add-members: add members to group
+  ::
+  ::    no-op if group does not exist
```

Comment:
Shouldn't it be pretty loud about this though? Silent failures aren't very nice...

Reply:
I agree with this, I think the only time you want to be silent about failures is if being loud gives away information you don't want to give away. But since only our ship can send this poke, we should go loud.

Reply:
This still remains, I think.